### PR TITLE
Dockerfile

### PR DESCRIPTION
**Pull request by kelvich**
Copied from Dockerfile comment:

```
# Docker image for console integration testing.
#
# We may also reuse it in CI to unify installation process and as a general binaries building
# tool for production servers.
#
# Dynamic linking is used for librocksdb and libstdc++ bacause librocksdb-sys calls
# bindgen with "dynamic" feature flag. This also prevents usage of dockerhub alpine-rust
# images which are statically linked and have guards against any dlopen. I would rather
# prefer all static binaries so we may change the way librocksdb-sys builds or wait until
# we will have our own storage and drop rockdb dependency.
#
# Cargo-chef is used to separate dependencies building from main binaries building. This
# way `docker build` will download and install dependencies only of there are changes to
# out Cargo.toml files.
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# images which are statically linked and have guards against any dlopen. I would rather
+# prefer all static binaries so we may change the way librocksdb-sys builds or wait until
+# we will have our own storage and drop rockdb dependency.
+#
+
+#
+# build postgres separately -- this layer will be rebuilt only if one of
+# mentioned paths will get any changes
+#
+FROM alpine:3.13 as pg-build
```
Comment (ololobus) [inline Dockerfile:18]: What is the reason of using two different alpine tags, ‘alpine:20210212’ and ‘alpine:3.13’? It looks like you did it intentionally, but it is not clear for me why

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#
+# Copy binaries to resulting image.
+# build-base hare to provide libstdc++ (it will also bring gcc, but leave it this way until we figure
+# out how to statically link rocksdb or avoid at all).
+#
+FROM alpine:3.13
+RUN apk add --update openssl build-base
+RUN apk --no-cache --update --repository https://dl-cdn.alpinelinux.org/alpine/edge/testing add rocksdb
+WORKDIR zenith
+COPY --from=build /zenith/target/release/pageserver /usr/local/bin
```
Comment (funbringer) [inline Dockerfile:79]: Some steps are missing, cause I can't start the pageserver:

```
/usr/local/bin # ./pageserver 
Error: Can't find postgres binary at "/usr/local/bin/tmp_install"
```
Then, after I override `export POSTGRES_DISTRIB_DIR=/usr/local`:

```
/usr/local/bin # ./pageserver 
Error: No such file or directory (os error 2)
```

And even with `LD_LIBRARY_PATH=/usr/lib` (btw why would I need to set this up?).

```
/usr/local/bin # strace -f -Z ./pageserver 
open("/usr/lib/libz.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/local/bin/../../../../usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libz.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/rustlib/x86_64-alpine-linux-musl/lib/libz.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/etc/ld-musl-x86_64.path", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
chdir(".zenith")                        = -1 ENOENT (No such file or directory)
Error: No such file or directory (os error 2)
+++ exited with 1 +++
```

Comment (funbringer) [reply Dockerfile:79]: I gave up early, but we'd better get rid of those paper cuts.

File: Dockerfile (hunk 1 above)
Comment (kelvich) [reply Dockerfile:18]: I've used edge alpine in cases where I bring rocksdb package which is for edge. It seems to work okay on 3.13, but just to decrease amount of potential problems. However I accidentally used 3.13 instead of edge in resulting package -- will fix.

File: Dockerfile (hunk 2 above)
Comment (kelvich) [reply Dockerfile:79]: Yeah, I'm writing right now proper entrypoint.sh which will set all env variables and init pageserver. And in passing I may add some missing libraries from resulting container (if any) + fix work with env in pageserver itself.

File: control_plane/Cargo.toml [hunk 3]
```diff
@@ -16,11 +16,8 @@ toml = "0.5"
 lazy_static = "1.4"
 regex = "1"
 anyhow = "1.0"
-# hex = "0.4.3"
 bytes = "1.0.1"
-# fs_extra = "1.2.0"
 nix = "0.20"
-# thiserror = "1"
```
Comment (funbringer) [inline control_plane/Cargo.toml:23]: Are those deps a dead weight? If so, I'd rather have them removed in a separate patch.

File: pageserver/Cargo.toml [hunk 4]
```diff
@@ -30,7 +30,8 @@ tokio-stream = { version = "0.1.4" }
 postgres-types = { git = "https://github.com/zenithdb/rust-postgres.git", rev="9eb0dbfbeb6a6c1b79099b9f7ae4a8c021877858" }
 postgres-protocol = { git = "https://github.com/zenithdb/rust-postgres.git", rev="9eb0dbfbeb6a6c1b79099b9f7ae4a8c021877858" }
 postgres = { git = "https://github.com/zenithdb/rust-postgres.git", rev="9eb0dbfbeb6a6c1b79099b9f7ae4a8c021877858" }
-rocksdb = "0.16.0"
+# by default rust-rocksdb tries to build a lot of compression algos. Use lz4 only for now as it is simplest dependency.
+rocksdb = { version = "0.16.0", features = ["lz4"], default-features = false }
```
Comment (funbringer) [inline pageserver/Cargo.toml:34]: This is also a notable change which deserves its own patch. I totally get that it's inconvenient, but it's so much better for git history.

File: control_plane/Cargo.toml (hunk 3 above)
Comment (kelvich) [reply control_plane/Cargo.toml:23]: that diff in separate commit already

File: pageserver/Cargo.toml (hunk 4 above)
Comment (kelvich) [reply pageserver/Cargo.toml:34]: 👌

**Comment by ericseppanen** [comment]
My preference would be to build the code exactly once in CI, so that the binary that the tests run against is exactly the same binary that gets deployed to the cloud.

I know I haven't done it yet, but if the CircleCI build job outputs a binary artifact (e.g. tarball), could that be loaded into the image instead of re-compiling?

I've also seen many reports that Rust binaries have performance problems when linked against musl-- I would suggest starting from debian/ubuntu docker images rather than alpine, until we feel like we have enough people to be monitoring performance on multiple targets.

**Comment by kelvich** [comment]
> My preference would be to build the code exactly once in CI, so that the binary that the tests run against is exactly the same binary that gets deployed to the cloud.
> 
> I know I haven't done it yet, but if the CircleCI build job outputs a binary artifact (e.g. tarball), could that be loaded into the image instead of re-compiling?

We would lose then ability to locally build images/binaries. Necessity to make an ad-hoc operation will happen and we need a clear way to do it (e.g. when we spotted some bug for some client and need to investigate things and put custom  binary for that situation). But as for general release pipeline we should use CI artifacts for sure. And we can use same dockerfile to produce binaries.

> I've also seen many reports that Rust binaries have performance problems when linked against musl-- I would suggest starting from debian/ubuntu docker images rather than alpine, until we feel like we have enough people to be monitoring performance on multiple targets.

Thanks for bringing this up. I've google around rust+musl and in both cases that i've found performance regression was due to malloc used in by default in musl at the time of writing this articles. Right now different malloc implementation is used in musl, so that is slightly outdated. And it would be anyway good idea to check performance with different allocators.

So I'm going to commit that as it is, but will do perf tests of musl vs. libc when we will get closer to the actual benchmarking.
